### PR TITLE
New Lazy Traffic Controller!

### PR DESCRIPTION
**Pull request by indraneelpatil**
1. Added velocity obstacles as a header only library
2. New controller which uses above library to calculate velocities
3. Added tests  for the above
4. Added executable node to run the above

### CODE REVIEW

File: test/simple_test.cpp [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#include <gtest/gtest.h>
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply test/simple_test.cpp:1]: Yeah need this in every test file

File: src/lazy_traffic_agent.cpp [hunk 2]
```diff
@@ -0,0 +1,220 @@
+// Created by Indraneel on 22/09/22
+
+#include "lazy_traffic_agent.hpp"
+
+#define PI 3.14159265
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply src/lazy_traffic_agent.cpp:5]: Done

File: include/lazy_traffic_rvo.hpp [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        else {
+            dist_to_pref_vel = abs(vel_cand - vel_pref);
+            dist_to_cur_vel = abs(vel_cand - vel_curr);
+        }
+
+        // searching for smallest time to collision with this velocity sample
+        float min_t_to_collision = RVO_INFTY;
+        
+        // iterate over neighbors
+        for(auto n: neighbors_list) {
```
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply include/lazy_traffic_rvo.hpp:120]: Done

File: src/lazy_traffic_agent.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (indraneelpatil) [reply src/lazy_traffic_agent.cpp:5]: Done

File: include/lazy_traffic_agent.hpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+#include "robosar_messages/controller_status.h"
+
+#include "Vector2.h"
+#include "lazy_traffic_rvo.hpp"
+
+typedef pair<string, float> AgentDistPair;
+
+using namespace std;
+
+#define MAX_NEIGHBORS 4 // Maximum number of neighbors to consider
```
[the remark this answers is not in the document]
Comment (stark710) [reply include/lazy_traffic_agent.hpp]: Resolved

File: include/lazy_traffic_agent.hpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    void computeNearestNeighbors(std::unordered_map<std::string, Agent> agent_map);
+    RVO::Vector2 getCurrentHeading();
+    
+    ros::Publisher pub_vel_;
+    ros::Publisher pub_status_;
+    ros::NodeHandle nh_;
+    geometry_msgs::TransformStamped lookahead_;
+    robosar_messages::controller_status status;
+
+    double v_max_;
```
[the remark this answers is not in the document]
Comment (stark710) [reply include/lazy_traffic_agent.hpp:71]: Resolved

File: include/lazy_traffic_agent.hpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+#include <string>
+#include <queue>
+
+// ROS stuff
+#include <geometry_msgs/Twist.h>
+#include <geometry_msgs/PoseStamped.h>
+#include <geometry_msgs/TransformStamped.h>
+#include <ros/ros.h>
+#include <tf/tf.h>
```
[the remark this answers is not in the document]
Comment (stark710) [reply include/lazy_traffic_agent.hpp:14]: Resolved

File: include/lazy_traffic_controller.hpp [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+#include <queue>
+#include <thread>
+#include <mutex>
+#include<unordered_map>
+
+#include "robosar_messages/robosar_controller.h"
+#include "lazy_traffic_agent.hpp"
+
+// ROS stuff
+#include <tf/tf.h>
```
[the remark this answers is not in the document]
Comment (stark710) [reply include/lazy_traffic_controller.hpp:17]: Resolved

File: include/lazy_traffic_rvo.hpp [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+
+//A class that operates on 2D
+#include "Vector2.h"
+#include <queue>
+#include <map>
+#include <unordered_map>
+#include <ros/console.h>
+
+#define RVO_VELOCITY_SAMPLES 1000 //NUMBER OF SAMPLES PER EACH AGENT
```
[the remark this answers is not in the document]
Comment (stark710) [reply include/lazy_traffic_rvo.hpp]: Resolved

File: include/lazy_traffic_rvo.hpp [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    return fabs(a - b) < std::numeric_limits<double>::epsilon();
+}
+
+//Function to compute if agent is in collision
+  inline float rvoTimeToCollision(const RVO::Vector2& p, const RVO::Vector2& v,
+                         const RVO::Vector2& p2, float radius, bool collision) {
+
+    //ROS_INFO(" RVO received p1: %f, %f, p2: %f, %f, v: %f, %f r:%f", p.x(), p.y(), p2.x(), p2.y(), v.x(), v.y(),radius);
+    RVO::Vector2 ba = p2 - p;
+    float sq_diam = radius * radius;
```
[the remark this answers is not in the document]
Comment (stark710) [reply include/lazy_traffic_rvo.hpp:46]: Resolved

File: include/lazy_traffic_rvo.hpp [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+        }
+
+        float dist_to_pref_vel ; // distance between candidate velocity and preferred velocity
+        float dist_to_cur_vel ; // distance between candidate velocity and current velocity
+        if(is_collision) {
+            dist_to_pref_vel = 0;
+            dist_to_cur_vel = 0;
+        }
+        else {
+            dist_to_pref_vel = abs(vel_cand - vel_pref);
```
[the remark this answers is not in the document]
Comment (stark710) [reply include/lazy_traffic_rvo.hpp:111]: Resolved

File: include/lazy_traffic_rvo.hpp [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+#include <ros/console.h>
+
+#define RVO_VELOCITY_SAMPLES 1000 //NUMBER OF SAMPLES PER EACH AGENT
+#define RVO_AGENT_RADIUS 0.15 // Radius of agent
+#define RVO_RADIUS_MULT_FACTOR 2 // since r1 + r2 = 2*r for RoboSAR, Define agent_radius and mult_factor for obstacle cone
+#define TIME_STEP 1 //frequence at which controller runs ( 1/ timestep)
+#define RVO_SAFETY_FACTOR 20.0f //The safety factor of the agent (weight for penalizing candidate velocities - the higher the safety factor, the less 'aggressive' an agent is)
+#define RVO_INFTY 9e9f
+
+using namespace std;
```
[the remark this answers is not in the document]
Comment (stark710) [reply include/lazy_traffic_rvo.hpp]: Resolved

File: include/lazy_traffic_rvo.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (stark710) [reply include/lazy_traffic_rvo.hpp]: Resolved

File: src/lazy_traffic_agent.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (stark710) [reply src/lazy_traffic_agent.cpp:5]: Resolved

**Pull request merged**